### PR TITLE
Possible fix for issue #437

### PR DESCRIPTION
**Pull request by SNiLD**
Probably my solution doesn't perfectly fit what you had planned for but this worked in my environment (Ogre 1.10.9).

### CODE REVIEW

**Comment by paroj** [comment]
how about putting `mGLUniformReferences` into the respective {fragment, vertex, geom} programs where they would be cleared on unload()?

I think this idea had some issues, but I do not recall them any more.. :confused:

**Comment by paroj** [comment]
the underlying issue is probably in `GLSLLinkProgramManager::setActiveXXXShader` which only checks the pointer but not the program status..

**Comment by SNiLD** [comment]
Yes. The program definitely has to be re-linked and re-created since the old programs have been destroyed as well.

**Comment by paroj** [comment]
you solution is definitely the right approach, I just think it is at the wrong place.

It basically duplicates this code:
https://github.com/OGRECave/ogre/blob/73928a19d70966a6c6c18624b6a7034ba6b559da/RenderSystems/GL/src/GLSL/src/OgreGLSLLinkProgramManager.cpp#L111-L124

which seems not to be working correctly - so we rather should fix the code above,

**Comment by SNiLD** [comment]
Shader ID stays the same because it is only assigned during the construction of the object. GLHandle however changes depending on the unloading/loading. If the code you referred was changed to use GLHandle instead it would find/create new programs but there would be no way to find the old program and destroy it (which would lead to memory leaks).

**Comment by paroj** [comment]
The code is already broken, as it leaks if you remove the HighLevelGpuProgram instead of just reloading it. I would rather  fix one issue at a time and consider the leaking a separate issue.

**Comment by SNiLD** [comment]
Sounds like HighLevelGpuProgram needs mechanism to notify listeners that it will be removed/loaded/unloaded and GLSLLinkProgramManager needs to lister for these. Doesn't the ResourceManager provide this type of functionality? That would fix both of these issues with one go (memory leak and unload/reload).

**Comment by paroj** [comment]
actually, yes
https://github.com/OGRECave/ogre/blob/6c2c08fb5b609e1f2ee9deb771348ec53fa78394/OgreMain/include/OgreResource.h#L423

**Comment by SNiLD** [comment]
Nice. Except it doesn't provide information when the resource is deleted. However this probably could be done through GLSLGpuProgram destructor. It could notify GLSLLinkProgramManager that the `mGLSLProgram` is no longer valid and GLSLLinkProgramManager should delete the appropriate GLSLLinkProgram.

**Comment by paroj** [comment]
actually we can always delete the program on unload (which also happens on deletion), as the program has to be equally re-created in both cases.

**Comment by SNiLD** [comment]
Maybe GLSLProgram should override the `GLSLShaderCommon::unloadImpl()` and first tell GLSLLinkProgramManager that the resource will be gone and the reset the shared pointer.

**Comment by paroj** [comment]
the problem is that GLSLProgram (shader actually) knows nothing of its parent Program. Also the solution should scale to GL3+ where we have separable programs (a shader can be part of multiple programs).
Therefore I prefer the listener approach. It might be slower but this is neglectable compared to the incurred shader compilation.

**Comment by SNiLD** [comment]
Actually `GLSLGpuProgram::unloadImpl()` should probably notify GLSLLinkProgramManager as it already notifies on bindProgram and unbindProgram. Problem is that once one GLSLGpuProgram is gone, it might affect many GLSLLinkPrograms and all of the should be searched for removal.

**Comment by SNiLD** [comment]
This new method for removal should also take care of the possible memory leaks. The removal method is bit brutal since it's brute force but I'm under the impression that unloads are not that common? This way at least the code stays simple.

Btw. for some reason the diff shows modifications from my previous commit as well although I've undone them.

**Comment by SNiLD** [comment]
I guess same issue persists on GL3+ side so that should include similar fix.

File: RenderSystems/GL/src/GLSL/src/OgreGLSLLinkProgramManager.cpp [hunk 1]
```diff
@@ -396,5 +396,30 @@ namespace Ogre {
...
+                iLinkProgram->second->getGeometryProgram() == gpuProgram)
+            {
+                if (mActiveLinkProgram == iLinkProgram->second)
+                {
+                    mActiveLinkProgram = NULL;
+                    // change back to fixed pipeline
+                    glUseProgramObjectARB(0);
+                }
+                delete iLinkProgram->second;
+                iLinkProgram = mLinkPrograms.erase(iLinkProgram);
```
Comment (paroj) [inline RenderSystems/GL/src/GLSL/src/OgreGLSLLinkProgramManager.cpp]: https://stackoverflow.com/questions/263945/what-happens-if-you-call-erase-on-a-map-element-while-iterating-from-begin-to

**Comment by paroj** [comment]
fun fact: this is already implemented for GLES2:
https://github.com/OGRECave/ogre/blob/a3d7681b0fad5549138af57746a6cd8fcc9d5fd9/RenderSystems/GLES2/src/GLSLES/src/OgreGLSLESLinkProgramManager.cpp#L145-L163

**Comment by SNiLD** [comment]
Hey, unifying the shader stuff is the right way to go. Should I close this pull request?

**Comment by SNiLD** [comment]
Also kind of related is that `mUniformRefsBuilt` in `GLSLLinkProgram` shadows the `mUniformRefsBuilt` of `GLSLProgramCommon`. Which I had resolved in my commit but don't see in #577 .

**Comment by paroj** [comment]
yes, lets move further discussion to #577

**Pull request closed without merging**